modules/core/src/main/java/org/apache/ignite/internal/processors/cache/warmup/LoadAllWarmUp.java

```diff
@@ -0,0 +1,207 @@
+                                + grp.cacheOrGroupName() + ", partCnt=" + (i + 1) + "partPageCnt=" + j +
+                                ", grpPageCnt=" + grpPageCnt + ", allPartCnt=" + allPartCnt + ", allPageCnt=" +
+                                allPageCnt + ']');
+                        }
+
+                        return;
+                    }
+
+                    long pagePtr = -1;
+
+                    try {
+                        pagePtr = pageMemEx.acquirePage(grp.groupId(), pageId);
+                    }
+                    finally {
+                        if (pagePtr != -1)
+                            pageMemEx.releasePage(grp.groupId(), pageId, pagePtr);
+                    }
+                }
+            }
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public void stop() throws IgniteCheckedException {
+        stop = true;
+    }
+
+    /** {@inheritDoc} */
+    @Override public String toString() {
+        return S.toString(LoadAllWarmUp.class, this);
+    }
+
+    /**
+     * Getting count of pages available for loading into data region.
+     *
+     * @param region Data region.
+     * @return Count(non-negative) of pages available for loading into data region.
+     */
+    protected long availableLoadPageCount(DataRegion region) {
+        long maxSize = region.config().getMaxSize();
+        long curSize = region.pageMemory().loadedPages() * region.pageMemory().systemPageSize();
+
+        return Math.max(0, (maxSize - curSize) / region.pageMemory().systemPageSize());
+    }
+
+    /**
+     * Calculation of cache groups, count of partitions and pages that can load
+     * into data region. Calculation starts and includes an index partition for
+     * each group.
+     *
+     * @param region Data region.
+     * @return Mapping: {cacheGrp -> {partCnt, pageCnt}}.
+     * @throws IgniteCheckedException – if faild.
+     */
+    protected Map<CacheGroupContext, T2<Integer, Long>> loadDataInfo(
+        DataRegion region
+    ) throws IgniteCheckedException {
+        // Get cache groups of data region.
+        List<CacheGroupContext> regionGrps = grpCtxSup.get().stream()
+            .filter(grpCtx -> region.equals(grpCtx.dataRegion())).collect(toList());
+
+        long availableLoadPageCnt = availableLoadPageCount(region);
+
+        // Computing groups, partitions, and pages to load into data region.
+        Map<CacheGroupContext, T2<Integer, Long>> loadableGrps = new LinkedHashMap<>();
+
+        for (int i = 0; i < regionGrps.size() && availableLoadPageCnt > 0; i++) {
+            CacheGroupContext grp = regionGrps.get(i);
+
+            int partCnt = 0;
+            long pageCnt = 0;
+
+            // Index partition in priority.
+            for (int j = -1; j < grp.topology().localPartitions().size() && availableLoadPageCnt > 0; j++, partCnt++) {
+                int partPageCnt = grp.shared().pageStore().pages(grp.groupId(), (j == -1 ? INDEX_PARTITION : j));
```

Comment:
Do you sure we don't want to load partitions only after all indexes from all groups would be loaded?